apps/src/util/useFetch.js

```diff
@@ -0,0 +1,93 @@
+  loading: false,
+  data: null,
+  error: null
+};
+
+/**
+ * React hook for making a fetch request to a resource that returns a JSON
+ * response.
+ *
+ * By default, fetch requests are configured with the following options:
+ * - credentials: 'same-origin'
+ * The default options can be overriden by the passed-in options.
+ *
+ * The 'loading' field of the return value indicates whether there is currently
+ * a request in flight. Note that this value will initially be false and then
+ * transition to true as a request is sent and internal state is updated
+ * asynchronously.
+ *
+ * If the fetch request is successful, the body is parsed as json and the
+ * resulting object is put in the 'data' field of the return value. If the
+ * request is not successful or the body could not be parsed as json, the
+ * 'error' field will contain an object representing the error.
+ *
+ * @param {string} url - URL of resource to fetch
+ * @param {RequestInit} options - options to pass to fetch
+ * @param {any[]} deps - array of values that fetch depends on; a new fetch
+ *    request will be sent if any of the values in the array change
+ * @returns {{loading: boolean, data: Object, error: Object}}
+ */
+export const useFetch = (url, options, deps) => {
+  const [fetchState, setFetchState] = useState(baseFetchState);
+
+  useEffect(() => {
+    // This local variable tracks whether this instance of the effect has been
+    // cancelled. One important case where this is true is when a previous instance
+    // of the effect is canceled before a new instance is created when deps change.
+    let canceled = false;
+
+    // Likewise, the abortController is also a local variable and specifc to
+    // this instance of the effect.
+    let abortController;
+    if (window.AbortController) {
+      abortController = new AbortController();
+    }
+
+    options = {
+      credentials: 'same-origin',
+      signal: abortController ? abortController.signal : undefined,
+      ...options
+    };
+
+    // This async IIFE contains the core of the fetch logic. Note that the
+    // setFetchState calls are asynchronous so state that is set here isn't
+    // returned to the caller until (at least) the next render cycle.
+    (async () => {
+      setFetchState({...baseFetchState, loading: true});
+      try {
+        const response = await fetch(url, options);
+        if (!response.ok) {
+          throw new Error(
+            `fetch request to ${url} failed with status code ${response.status}`
+          );
+        }
+
+        const data = await response.json();
+        if (!canceled) {
+          setFetchState({...baseFetchState, data: data});
+        }
+      } catch (e) {
+        console.error(e);
+        if (!canceled) {
+          setFetchState({...baseFetchState, error: e});
+        }
+      }
+    })();
+
+    return () => {
+      if (abortController) {
+        canceled = true;
+        abortController.abort();
+      }
+    };
+  }, deps);
```

Comment:
I tried this originally, but I ended up in a rendering loop.  I think it was because options is an object and deps uses object equality?  Should I serialize options to JSON or something like that?

Reply:
I think you should be able to use useMemo in the component that renders this hook to keep a reference to the object static, if we have const options = {} somewhere, that'll create a new reference each time

Reply:
Oh, I see.  It seems nice for the caller to not have to pass deps but having to remember to use `useMemo` also seems like a stumbling block.

Related question, do you think this hook should support scenarios where the caller might want to re-fetch even though the url and options haven't changed?

Reply:
I think in that case, i.e. sending analytics or something, we would want something other than a hook, or something that better controls the request rate. By using a hook, we're also implicitly deciding that the api isn't going to change between calls.

Reply:
Re useMemo, this is also the case for several other hooks api things like useEffect. Its a little inconvenient, but that comes with hooks. There was a comment in the hooks docs about react potentially being able to automate the deps array in future versions, but its not quite there yet.

Reply:
I've taken your suggestion, thanks!